### PR TITLE
TestManager, cli, Agent: Improvements to `status`

### LINKED ISSUE #658: agent: Consider adding a last update field to crd status
The status command could show the last time the crd was updated.

When calling `testsys status` it may be useful to have the last time the crd was updated displayed i.e 
```
 NAME                    TYPE      STATE      PASSED      SKIPPED      FAILED     LAST UPDATED
 x86-64-aws-k      Test       Pass          1                   0                0         17:26:38Z
```
or
```
 NAME                    TYPE      STATE      PASSED      SKIPPED      FAILED     TIME SINCE UPDATE
 x86-64-aws-k      Test       Pass         1                   0                0           540s
```

### PR DESCRIPTION
**Pull request by ecpullen**
**Issue number:**

Related to #679 
Closes #658 

**Description of changes:**

```
commit 75e173d2fa131244823cb337731cd2142a4f136d (HEAD -> testsys-status, origin/testsys-status)
Author: ecpullen <ecpullen@aol.com>
Date:   Mon Dec 5 18:41:17 2022 +0000

    TestManager: Don't truncate `Last Update`
    
    This sets the `MinWidth` for the `LAST UPDATE` column to 20 chars (the
    exact number required).

commit 318c88c5ccdb3430ead02e1295e087acff08b3e4
Author: ecpullen <ecpullen@aol.com>
Date:   Mon Dec 5 18:28:28 2022 +0000

    TestManager: Fix some columns out of order
    
    When `with_progress` and `with_time` were added, the order of crds was
    allowed to change. This caused the columns to be out of order. The crds
    are now resorted in each function.

commit cc26f55cb4a7a213054cf6f74776823648edb614
Author: ecpullen <ecpullen@aol.com>
Date:   Mon Dec 5 17:59:02 2022 +0000

    Controller, Agent: Report the last time of update
    
    Adds a new field to both `Test` and `Resource` status.
    All json patches now add `new_timestamp()` which adds the `SystemTime`
    to the `status/lastUpdate`.

commit ae036a83496dff45b02e97de419912a36775b2b5
Author: ecpullen <ecpullen@aol.com>
Date:   Fri Dec 2 18:26:44 2022 +0000

    TestManager: Additional column duplication
    
    This fixes a bug where a test with retries would cause additional column
    to not be aligned because the values were not duplicated for each result
    row.

commit d50efaac1d70f5802929553e2fdaee1b2a4225e7
Author: ecpullen <ecpullen@aol.com>
Date:   Fri Dec 2 18:07:39 2022 +0000

    TestManager: Add optional `STATUS` column
    
    A new optonal status column was added that displays the `currentStatus`
    of a resource or optionally the `other_info` of a test.
    `StatusProgress::Resources` will only show the status of resources while
    `StatusProgress::WithTests` will also include test status.
```

**Testing done:**

Tested with `cargo make test`

Initial status
```bash
$ cli status -pu
 NAME                                TYPE       STATE     PASSED   SKIPPED   FAILED  STATUS            LAST UPDATE           
 x86-64-aws-k8s-123                  Resource   unknown                                                 2022-12-05T18:47:44Z 
 x86-64-aws-k8s-123-instances-quic   Resource   unknown                                                 2022-12-05T18:47:44Z 
 x86-64-aws-k8s-123-test             Test       waiting                               No test results   2022-12-05T18:47:44Z 
```

After cluster creation
```bash
$ cli status -pu
 NAME                                TYPE      STATE      PASSE   SKIPPE   FAILE  STATUS               LAST UPDATE           
 x86-64-aws-k8s-123                  Resourc   complete                            Cluster ready        2022-12-05T18:47:47Z 
 x86-64-aws-k8s-123-instances-quic   Resourc   complete                            Instance(s) Create   2022-12-05T18:47:55Z 
 x86-64-aws-k8s-123-test             Test      running                             No test results      2022-12-05T18:48:00Z 
```

After testing is complete
```bash
$ cli status -pu
 NAME                                TYPE      STATE      PASSE   SKIPPE   FAILED  STATUS              LAST UPDATE           
 x86-64-aws-k8s-123                  Resourc   complete                             Cluster ready       2022-12-05T18:47:47Z 
 x86-64-aws-k8s-123-instances-quic   Resourc   running                              Instances deleted   2022-12-05T18:48:47Z 
 x86-64-aws-k8s-123-test             Test      passed     1       7051     0                            2022-12-05T18:48:36Z 
```

`with-tests`
```bash
$ cli status -put
 NAME                            TYP   STATE   PA   SKI   FA  STATUS                                       LAST UPDATE       
 x86-64-aws-k8s-123              Res   compl                   Cluster ready                                2022-12-05T18:47 
 x86-64-aws-k8s-123-instances-   Res   runni                   Instances deleted                            2022-12-05T18:48 
 x86-64-aws-k8s-123-test         Tes   passe   1    705   0    {"name":"e2e","node":"global","timestamp":   2022-12-05T18:48 
```

Without any flags (unchanged by this pr)
```
NAME                               TYPE               STATE               PASSED           SKIPPED           FAILED         
 x86-64-aws-k8s-123                 Resource           completed                                                             
 x86-64-aws-k8s-123-test            Test               passed              1                7051              0              
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Review by webern — APPROVED** [review]
Nice.

**Comment by mjsterckx** [comment]
Since the content of the `STATE` column is lowercase, can we do the same for `STATUS`?

**Comment by ecpullen** [comment]
> Since the content of the `STATE` column is lowercase, can we do the same for `STATUS`?

All headers are currently caps.

**Comment by ecpullen** [comment]
> Since the content of the `STATE` column is lowercase, can we do the same for `STATUS`?

This status is whatever is the agent reports. As a follow up to this PR, I will be expanding the frequency of status updates in the agents.

**Comment by mjsterckx** [comment]
> > Since the content of the `STATE` column is lowercase, can we do the same for `STATUS`?
> 
> All headers are currently caps.

I mean `running`, `completed`, etc. in `STATE` are all lowercase, but `Cluster ready` in `STATUS` for example is capitalized.

**Comment by mjsterckx** [comment]
Just read your second reply. Sounds good.

**Review by mjsterckx — APPROVED** [review]
LGTM

File: cli/src/status.rs [hunk 1]
```diff
@@ -13,15 +13,37 @@ pub(crate) struct Status {
...
+    #[clap(long, short = 'p')]
+    progress: bool,
+
+    /// Include the `Test` status, too. Requires `--progress`
+    #[clap(long, short = 't', requires("progress"))]
+    with_test: bool,
+
+    /// Include the time the CRD was last updated
+    #[clap(long, short = 'u')]
+    with_time: bool,
```
Comment (etungsten) [inline cli/src/status.rs:27]: The options here are kind of hard to follow. `--progress/-p` shows status for resources, but not tests, for that you need `--with-test/-t`. If you want the CRD update time, you have to include `-u` which doesn't seem that intuitive.

I wonder if we could adopt similar to `kubectl get`'s `-o` option where `json` means json formatted output and `wide` means everything in a table. For custom tables they allow it through https://kubernetes.io/docs/reference/kubectl/#custom-columns. 

Let's not block on implementing that but I think we should just output the test object status by default since we care about that the most when kicking off tests and include a `--wide` option that outputs everything. So we'd end up with just three output options for now, `--json`, `--wide` and the default output.

**Review by etungsten — APPROVED** [review]

File: cli/src/status.rs (hunk 1 above)
Comment (stmcginnis) [reply cli/src/status.rs:27]: I like that pattern!

**Review by stmcginnis — APPROVED** [review]
